posts/inside-rust/2020-02-07-intro-rustc-self-profile.md

```diff
@@ -0,0 +1,269 @@
+
+1. Open Chrome
+2. Open the Developer Tools console by pressing `Ctrl` + `Shift` + `i` (Windows/Linux) or `Cmd` + `Option` + `i` (macOS)
+3. Click the Performance tab at the top of the console.
+4. Click the "Load profile" button which looks like an arrow pointing up.
+5. Select the `chrome_profiler.json` file we crated.
+
+You should now see something similar to this:
+
+![Image of chrome profiler][chrome profiler img1]
+
+You can use the scroll wheel on a mouse or the appropriate gesture on a touchpad to zoom in or out of the timeline.
+
+### Filtering short events
+
+If the `chrome_profiler.json` file gets too large, the normal Chromium performance tools have issues opening the file.
+One easy way to deal with this is to tell `crox` to remove events shorter than a chosen duration:
+
+```sh
+$ ../measureme/target/release/crox --minumum-duration 2 regex-17088
+```
+
+Filtering out events less than 2 microseconds shrinks our `chrome_profiler.js` file from 27mb to 11mb.
+
+### Capturing event arguments
+
+The self-profiler can be configured to record event arguments during compilation.
+For example, queries will include their query key.
+This fuctionality is turned off by default because it significantly increases the self-profiler overhead.
+
+To turn this feature on, we'll need to record a new compilation, passing an additional argument to `rustc`:
+
+```sh
+$ cargo clean
+$ cargo rustc -- -Zself-profile -Zself-profile-events=default,args
+```
+
+And then process the new output files:
+
+```sh
+$ ../measureme/target/release/crox regex-23649
+```
+
+Now in the Chromium profiler, if you click on a node, you can see additional data about many of the events at the bottom of the screen:
+
+![Image of Chrome profiler details][chrome profiler img2]
+
+Which shows this `optimized_mir` query was processing the `regex::compile::{{impl}}::new` function body.
+
+### Profiling an entire crate graph
+
+By using the `RUSTFLAGS` environment variable, we can profile every `rustc` invocation, not just the final crate's.
+`crox` can then combine all of the profiles together into one output file.
+
+```sh
+$ rm regex-17088.* regex-23649.*
+$ cargo clean
+$ RUSTFLAGS="-Zself-profile=$(pwd) -Zself-profile-events=default,args" cargo build
+```
+
+This creates quite a few trace files in the working directory.
+Now, we'll tell `crox` to combine all of the trace files in the current directory together:
+
+```sh
+$ ../measureme/target/release/crox --dir .
+```
+
+Opening this file shows all of the crates compiled:
+
+![Image of Chrome profiler with all crates][chrome profiler img3]
+
+Clicing on a crate will expand it to show the threads and event data inside it:
+
+![Image of Chrome profiler with a crate expanded][chrome profiler img4]
+
+Thanks for reading!
+If you have questions or would like to get involved with the Self-Profile Working Group, please check out the [measureme repository] or stop by our [Zulip stream].
```

Comment:
Maybe add some kind of summary statement like "We've been using these tools extensively ourselves over the last few months and they've helped us tremendously in understanding where the compiler spends its time. In the future we'll be adding more features and we'll work on making the tooling easier to use." 

Something like that..